app/controllers/show_product_controller.rb

```diff
@@ -0,0 +1,75 @@
+class ShowProductController < ApplicationController
+  def show
+    @product = Product.filtersearch(params[:search], params[:trademark],
+      params[:assort], params[:max], params[:min]).select :id, :name, :price,
+    :img_detail
+    if session[:cart_p].nil?
+      session[:cart_p] = {}
+    end
+    if @product.nil?
+      flash[:danger] = t("controllers.show_product.noproduct")
+      redirect_to showp_url
+    end
+  end
+
+  def detail
+    @product =  Product.left_outer_joins(:rates,:comments).
+    select("products.id, products.name, products.star, products.content,
```

Comment:
1:ở controller ko nên xử lý logic, toàn bộ hàm này em đưa xuống model sau đó gọi lên thông qua controller. 
2: để ngắn gọn hơn em thử tìm hiểu eager loading xem, a nghĩ có thể ngắn gọn hơn cách này đấy.

Reply:
Theo em tìm hiểu khi mình dùng eager_load thì code nó ngắn hơn thật nhưng theo ý anh muốn lấy count rate thì lại phải gọi thêm 1 câu sql mới. Em có làm VD và thấy thế, Em ko chắc cách em làm đúng ý anh bảo.
VD : p = Product.eager_load(:rates, :comments).find_by_id(1)
get count rate : p.rates.count
get count comment : p.comments.count